cgen: can't set a reference automatically for sum types.

### Describe the bug

```v
module main

struct ParExpr {
	expr Expr
}

struct InfixExpr {
	left  Expr
	right Expr
}

type Expr = InfixExpr | ParExpr

fn print_expr(expr &Expr) {
	println(expr)
}

fn main() {
	par := ParExpr{
		expr: InfixExpr{}
	}

	print_expr(par) // ok
	print_expr(Expr(par)) // error
	print_expr(par.expr) // error
	print_expr(&par.expr) // ok
}

```

### Expected Behavior

Successful compilation.

### Current Behavior

```
/tmp/v_501/cgen.2691344066624070828.tmp.c:12336:19: error: passing 'main__Expr' (aka 'struct main__Expr') to parameter of incompatible type 'main__Expr *' (aka 'struct main__Expr *')
        main__print_expr(main__ParExpr_to_sumtype_main__Expr(&par));
                         ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/tmp/v_501/cgen.2691344066624070828.tmp.c:12329:51: note: passing argument to parameter 'expr' here
VV_LOCAL_SYMBOL void main__print_expr(main__Expr* expr) {
                                                  ^
/tmp/v_501/cgen.2691344066624070828.tmp.c:12337:19: error: passing 'main__Expr' (aka 'struct main__Expr') to parameter of incompatible type 'main__Expr *' (aka 'struct main__Expr *'); take the address with &
        main__print_expr(par.expr);
                         ^~~~~~~~
                         &
/tmp/v_501/cgen.2691344066624070828.tmp.c:12329:51: note: passing argument to parameter 'expr' here
```

### Reproduction Steps

Run code from `Describe the bug`

### Possible Solution

_No response_

### Additional Information/Context

_No response_

### V version

V 0.3.2 bb512f7

### Environment details (OS name and version, etc.)

macOS 13.0

--- a/vlib/v/gen/c/fn.v
+++ b/vlib/v/gen/c/fn.v
@@ -2289,6 +2289,7 @@ fn (mut g Gen) keep_alive_call_postgen(node ast.CallExpr, tmp_cnt_save int) {
 [inline]
 fn (mut g Gen) ref_or_deref_arg(arg ast.CallArg, expected_type ast.Type, lang ast.Language) {
 	arg_typ := g.unwrap_generic(arg.typ)
+	arg_sym := g.table.sym(arg_typ)
 	exp_is_ptr := expected_type.is_ptr() || expected_type.idx() in ast.pointer_type_idxs
 	arg_is_ptr := arg_typ.is_ptr() || arg_typ.idx() in ast.pointer_type_idxs
 	if expected_type == 0 {
@@ -2300,7 +2301,6 @@ fn (mut g Gen) ref_or_deref_arg(arg ast.CallArg, expected_type ast.Type, lang as
 		g.write('&/*mut*/')
 	} else if exp_is_ptr && !arg_is_ptr {
 		if arg.is_mut {
-			arg_sym := g.table.sym(arg_typ)
 			if exp_sym.kind == .array {
 				if (arg.expr is ast.Ident && (arg.expr as ast.Ident).kind == .variable)
 					|| arg.expr is ast.SelectorExpr {
@@ -2352,6 +2352,19 @@ fn (mut g Gen) ref_or_deref_arg(arg ast.CallArg, expected_type ast.Type, lang as
 						needs_closing = true
 						g.write('ADDR(${g.typ(atype)}/*qq*/, ')
 					}
+				}
+			} else if arg_sym.kind == .sum_type && exp_sym.kind == .sum_type {
+				// Automatically passing sum types by reference if the argument expects it,
+				// not only the argument is mutable.
+				if arg.expr is ast.SelectorExpr {
+					g.write('&/*sum*/')
+					g.expr(arg.expr)
+					return
+				} else if arg.expr is ast.CastExpr {
+					g.write('ADDR(${g.typ(expected_deref_type)}/*sum*/, ')
+					g.expr_with_cast(arg.expr, arg_typ, expected_type)
+					g.write(')')
+					return
 				}
 			}
 		}

--- a/vlib/v/gen/c/testdata/sumtype_pass_by_reference.out
+++ b/vlib/v/gen/c/testdata/sumtype_pass_by_reference.out
@@ -1,0 +1,20 @@
+&Expr(ParExpr{
+    expr: Expr(InfixExpr{
+        left: unknown sum type value
+        right: unknown sum type value
+    })
+})
+&Expr(ParExpr{
+    expr: Expr(InfixExpr{
+        left: unknown sum type value
+        right: unknown sum type value
+    })
+})
+&Expr(InfixExpr{
+    left: unknown sum type value
+    right: unknown sum type value
+})
+&Expr(InfixExpr{
+    left: unknown sum type value
+    right: unknown sum type value
+})

--- a/vlib/v/gen/c/testdata/sumtype_pass_by_reference.vv
+++ b/vlib/v/gen/c/testdata/sumtype_pass_by_reference.vv
@@ -1,0 +1,27 @@
+module main
+
+struct ParExpr {
+	expr Expr
+}
+
+struct InfixExpr {
+	left  Expr
+	right Expr
+}
+
+type Expr = InfixExpr | ParExpr
+
+fn print_expr(expr &Expr) {
+	println(expr)
+}
+
+fn main() {
+	par := ParExpr{
+		expr: InfixExpr{}
+	}
+
+	print_expr(par)
+	print_expr(Expr(par))
+	print_expr(par.expr)
+	print_expr(&par.expr)
+}

